contracts/colony/ColonyExpenditure.sol

```diff
@@ -191,4 +229,46 @@ contract ColonyExpenditure is ColonyStorage {
   function getExpenditureSlotPayout(uint256 _id, uint256 _slot, address _token) public view returns (uint256) {
     return expenditureSlotPayouts[_id][_slot][_token];
   }
+
+  // Internal functions
+
+  bool constant MAPPING = false;
+  bool constant OFFSET = true;
+  uint256 constant MAX_OFFSET = 1024; // Prevent writing to arbitrary storage slots
+
+  function executeStateChange(
+    bytes32 _slot,
+    bool[] memory _mask,
+    bytes32[] memory _keys,
+    bytes32 _value
+  )
+    internal
+  {
+    require(_keys.length == _mask.length, "colony-expenditure-bad-mask");
+
+    bytes32 value = _value;
+    bytes32 slot = _slot;
+
+    // See https://solidity.readthedocs.io/en/v0.5.14/miscellaneous.html
+    for (uint256 i; i < _keys.length; i++) {
+
+      if (_mask[i] == MAPPING) {
+        slot = keccak256(abi.encode(_keys[i], slot));
+      }
+
+      if (_mask[i] == OFFSET) {
+        require(uint256(_keys[i]) <= MAX_OFFSET, "colony-expenditure-large-offset");
+
+        slot = bytes32(add(uint256(_keys[i]), uint256(slot)));
+        if (i != _keys.length - 1) { // If not last offset
```

Comment:
Let's add some explanation to this comment, something like
> If we are indexing in to an array, and this was the last entry in `keys`, then we have arrived at the storage slot that we want to set, and so do not hash the slot (which would take us to the start of the storage of a hypothetical array at this location).

Reply:
Sounds good.